pkg/e2e/autoscaler/autoscaler.go

```diff
@@ -176,166 +139,246 @@ func machineAutoscalerResource(targetMachineSet *mapiv1beta1.MachineSet, minRepl
-		o.Expect(err).NotTo(o.HaveOccurred())
-	}
+	testStartTime := metav1.Now()
 
-	scaleDownFunc := func(client runtimeclient.Client, targetMachineSet string, nodeTestLabel string, initialNumberOfReplicas int32) {
-		g.By(fmt.Sprintf("Waiting for machineset %q to have at most initial number of replicas", targetMachineSet))
-		err := wait.PollImmediate(5*time.Second, e2e.WaitLong, func() (bool, error) {
-			ms, err := e2e.GetMachineSet(context.TODO(), client, targetMachineSet)
-			if err != nil {
-				glog.Errorf("error getting machineset object: %v, retrying...", err)
-				return false, nil
+	eventInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			event := obj.(*corev1.Event)
+			if event.CreationTimestamp.Before(&testStartTime) {
+				return
 			}
-			msReplicas := pointer.Int32PtrDerefOr(ms.Spec.Replicas, e2e.DefaultMachineSetReplicas)
-			glog.Infof("Initial number of replicas: %d. Current number of replicas: %d", initialNumberOfReplicas, msReplicas)
-			if msReplicas > initialNumberOfReplicas {
-				return false, nil
+			if interestingComponents.Has(event.Source.Component) {
+				msg := fmt.Sprintf("%s/%s: %s\n", event.Source.Component, event.InvolvedObject.Name, event.Message)
+				g.GinkgoWriter.Write([]byte(msg))
 			}
+			matchEventFuncLock.Lock()
+			defer matchEventFuncLock.Unlock()
+			if matchEventFunc != nil {
+				matchEventFunc(event)
+			}
+		},
+	})
 
-			// Make sure all scaled down nodes are really gone (so they don't affect tests to be run next)
-			scaledNodes := corev1.NodeList{}
-			if err := client.List(context.TODO(), runtimeclient.MatchingLabels(map[string]string{nodeTestLabel: ""}), &scaledNodes); err != nil {
-				glog.Errorf("Error querying api for node objects: %v, retrying...", err)
-				return false, nil
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	kubeInformerFactory.Start(stopCh)
+	o.Expect(cache.WaitForCacheSync(stopCh, eventInformer.HasSynced)).Should(o.BeTrue())
+
+	var client runtimeclient.Client
+	client, err = e2e.LoadClient()
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	var cleanupObjects []runtime.Object
+	defer func() {
+		cascadeDelete := metav1.DeletePropagationForeground
+		for _, obj := range cleanupObjects {
+			if err := client.Delete(context.TODO(), obj, func(opt *runtimeclient.DeleteOptions) {
+				opt.PropagationPolicy = &cascadeDelete
+			}); err != nil {
+				glog.Errorf("error deleting object: %v", err)
 			}
-			scaledNodesLen := int32(len(scaledNodes.Items))
-			glog.Infof("Current number of replicas: %d. Current number of nodes: %d", msReplicas, scaledNodesLen)
-			return scaledNodesLen <= msReplicas && scaledNodesLen <= initialNumberOfReplicas, nil
-		})
+		}
+	}()
+
+	g.By("Getting machinesets")
+	machineSets, err := e2e.GetMachineSets(context.TODO(), client)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	// We need at least two machinesets to so that we can scale up
+	// to maxNodesTotal-1.
```

Comment:
we need two machinesets because min & max values in asr are such that (max - min == 1). If Max was let say 3, then even one machineset would have been sufficient to carry out the test. Right?